pkg/go/graph/weighted_graph.go

```diff
@@ -0,0 +1,399 @@
+package graph
+
+import (
+	"fmt"
+	"math"
+	"strings"
+)
+
+const Infinite = math.MaxInt32
+
+type WeightedAuthorizationModelGraph struct {
+	nodes map[string]*WeightedAuthorizationModelNode
+	edges map[string][]*WeightedAuthorizationModelEdge
+}
+
+// NewGraph creates a new Graph.
+func NewWeightedAuthorizationModelGraph() *WeightedAuthorizationModelGraph {
+	return &WeightedAuthorizationModelGraph{
+		nodes: make(map[string]*WeightedAuthorizationModelNode),
+		edges: make(map[string][]*WeightedAuthorizationModelEdge),
+	}
+}
+
+// AddNode adds a node to the graph with optional operationType and weight.
+func (wg *WeightedAuthorizationModelGraph) AddNode(uniqueLabel, label string, nodeType NodeType) {
+	wg.nodes[uniqueLabel] = &WeightedAuthorizationModelNode{uniqueLabel: uniqueLabel, label: label, nodeType: nodeType}
+}
+
+func (wg *WeightedAuthorizationModelGraph) AddEdge(fromID, toID string, edgeType EdgeType, condition string) {
+	fromNode := wg.nodes[fromID]
+	toNode := wg.nodes[toID]
+	edge := &WeightedAuthorizationModelEdge{from: fromNode, to: toNode, edgeType: edgeType, conditionedOn: condition}
+	wg.edges[fromID] = append(wg.edges[fromID], edge)
+}
+
+func (wg *WeightedAuthorizationModelGraph) AssignWeights() error {
+	visited := make(map[string]bool)
+	ancestorPath := make([]*WeightedAuthorizationModelEdge, 0)
+	tupleCycleDependencies := make(map[string][]*WeightedAuthorizationModelEdge)
+
+	for node := range wg.nodes {
+		if visited[node] {
+			continue
+		}
+
+		tupleCyles, err := wg.calculateNodeWeight(node, visited, ancestorPath, tupleCycleDependencies)
+		if err != nil {
+			return err
+		}
+		if len(tupleCyles) > 0 {
+			for _, tupleCycle := range tupleCyles {
```

Comment:
Don't think we need these logs as they'd be lost/invisible.

Reply:
ok, removing the logs